File: 21551-naveen-siva-kumar/week5/785-is-graph-bipartite/785-is-graph-bipartite.cpp
//This solution has passed  79/81 test cases    ------------using bfs
//have to use queues to do the same thing and see.



class Solution {
public:
    bool isBipartite(vector<vector<int>>& graph) {
        //int colors[] = new int[graph.size()]
        int colors[graph.size()];
        memset( colors, 0, graph.size()*sizeof(int) );
        
        // for(int i=0;i<graph.size();i++){
        //     cout<<colors[i]<<" ";
        // }
        // cout<<endl;

        int color = 0;
        cout<<" graph.size() :"<<graph.size()<<endl;
        for(int i=0;i<graph.size();i++){
            
            if(colors[i]==0){
                for(int k=0;k<graph[i].size();k++){
                    //cout<<"ima ";
                    if(colors[graph[i][k]]==1) {colors[i]=-1;color=-1;break;}
                    if(colors[graph[i][k]]==-1) {colors[i]=1;color=1;break;}
                }
                if(color==0) {colors[i]=1;color=1;}
                
            }
            if(colors[i]==1){
                color=1;
            }
            if(colors[i]==-1){
                color=-1;
            }
            cout<<"color :"<<color<<endl;
            for(int j=0;j<graph[i].size();j++){
                if(color==1){
                    //colors of neighbours should be -1
                    if(colors[graph[i][j]]==-1) continue;
                    if(colors[graph[i][j]]==1){ 
                      return false;
                    }
                    if(colors[graph[i][j]]==0) colors[graph[i][j]] =-1;
                    
                }
                if(color==-1){
                    //colors of neighbours should be -1
                    if(colors[graph[i][j]]==1) continue;
                    if(colors[graph[i][j]]==-1) return false;
                    if(colors[graph[i][j]]==0) colors[graph[i][j]] = 1;
                    
                }
            }
        }
        return true;
    
           
    }
};

File: 21561 - Suvam Tamang/Week 5/785_Is Graph Bipartite.java
class Solution {
    public boolean isBipartite(int[][] graph) {
        int[] colors = new int[graph.length];
        // 0 : Uncolored
        // 1 : Blue
        // -1: Red
        int color = 1;
        for(int i=0; i<graph.length; i++) {
            if(colors[i] == 0 && !dfsValidate(graph, colors, color, i))
                return false;
        }
        return true;
    }
    
    public Boolean dfsValidate(int[][] graph,int[] colors,int color,int node) {
        if(colors[node] != 0)
            return colors[node] == color;
        colors[node] = color;
        for(int next_node : graph[node]) {
            if(!dfsValidate(graph,colors,-color,next_node))
                return false;
        }
        return true;
    }
}
